client/lib/logmein/index.ts

```diff
@@ -0,0 +1,128 @@
+/**
+ * Internal Dependencies
+ */
+import { isEnabled } from '@automattic/calypso-config';
+import wpcom from 'calypso/lib/wp';
+
+/**
+ * Append logmein=1 query parameter to mapped domain urls we want the user to be logged in against.
+ */
+
+// Used as placeholder / default domain to detect when we're looking at a relative url
+const INVALID_URL = `https://__domain__.invalid`;
+
+export function logmeinUrl( url: string, sites: any, isWPComLoggedIn = true ): string {
+	let newurl: URL;
+
+	if ( ! isEnabled( 'logmein' ) ) {
+		return url;
+	}
+
+	try {
+		newurl = new URL( String( url ), INVALID_URL );
+	} catch ( e ) {
+		// Ignore unparseable urls
+		return url;
+	}
+
+	// Ignore and passthrough /relative/urls that have no host specified
+	if ( newurl.origin === INVALID_URL ) {
+		return url;
+	}
+
+	let allow = logmeinAllowedUrls( sites );
+	// Ignore urls not in the allow list
+	allow = allow.map( ( allowed: string ) => new URL( allowed, INVALID_URL ).hostname );
+	if ( allow.indexOf( newurl.hostname ) === -1 ) {
+		return url;
+	}
+
+	// logmein doesn't work with http.
+	newurl.protocol = 'https:';
+
+	// we're already logged into wordpress.com so we can take a shortcut
+	if ( isWPComLoggedIn ) {
+		newurl.searchParams.set( 'logmein', 'direct' );
+	} else {
+		newurl.searchParams.set( 'logmein', '1' );
+	}
+	return newurl.toString();
+}
+
+export function logmeinAllowedUrls( sites: any ): string[] {
+	return sites
+		.map( ( site: any ) =>
+			! site.is_vip &&
+			! site.jetpack &&
+			! site.options.is_automated_transfer &&
+			! site.options.is_domain_only &&
+			! site.options.is_redirect &&
+			! site.options.is_wpcom_atomic &&
+			! site.options.is_wpcom_store &&
+			! site.options.is_wpforteams_site &&
+			site.options.is_mapped_domain
+				? new URL( site.URL ).host
+				: false
+		)
+		.filter( Boolean );
+}
+
+export function appendLogmein( url: URL ): URL {
+	// logmein doesn't work with http.
+	url.protocol = 'https:';
+	url.searchParams.set( 'logmein', '1' );
+	return url;
+}
+export function appendLogmeinDirect( url: URL ): URL {
+	// logmein doesn't work with http.
+	url.protocol = 'https:';
+	url.searchParams.set( 'logmein', 'direct' );
+	return url;
+}
+
+let allowedSites: string[] = [];
+
+export function attachLogmein( isWPComLoggedIn = false ): void {
+	if ( ! isEnabled( 'logmein' ) ) {
+		return;
+	}
+
+	if ( isWPComLoggedIn ) {
+		wpcom
+			.me()
+			.sites()
+			.then( ( sites: any ) => {
+				allowedSites = logmeinAllowedUrls( sites.sites );
+				document.addEventListener( 'click', logmeinOnClick, false );
+			} );
+	}
+}
+
```

Comment:
This or the allow list may need to get more complicated:

The "View" post link loads the previewer using the `*.wordpress.com` domain, this is fine for the previewer but the previewer then links to this domain relying on redirects.

This code treats this situation as a no match and doesn't append the logmein param.

Fix probably involves doing a mapping step before the indexOf check.

Reply:
In that case you need to check `site.option.unmapped_url`, too. The `*.wordpress.com` variant is there. While `site.URL` contains the primary mapped domain.

Reply:
Editor won't work because its iframed to the *.wordpress.com site (Even on mapped domains because similar 3rd party cookie reasons) 

Fix needs to be implemented in [ETK](https://github.com/Automattic/wp-calypso/tree/trunk/apps/editing-toolkit) or move this to a package ETK can use. (why I started there first go round but it overcomplicated getting it working)

It can be done in a serverside plugin for simple sites only but I want this to work for atomic eventually so we need ETK for this.

Reply:
Thinking more: might be worth deploying this behind feature flag first and iterating on a solution with ETK, already a lot happening here. Potentially need to move a lot of files around though :thinking: let me see what an ETK fix might look like today.

Reply:
> The intercept also didn't work on the themes page, in the pop-up that appears after activating a theme:

That's because the "Visit site" is not a link, but a button that does `window.open( url )` in an `onClick` handler. We can't possibly intercept that, except maybe patching the `window.open` function. The best solution would be to convert that button to `<a href>`. That improves accessibility, too.

Reply:
> Editor won't work because its iframed to the *.wordpress.com site

One solution for that: intercept `<a href>` clicks in the editor iframe, and then send them to the parent window as an `openLink` message. The parent window has access to the sites list and can do the navigation properly.

I vaguely remember we do this technique already somewhere else -- maybe in post preview iframes? Previews do a lot of `postMessage` communication with their parent frame.

Reply:
> That's because the "Visit site" is not a link, but a button that does window.open( url ) in an onClick handler.

Didn't realize this was a button. Agree with changing to `<a>` but that will still require some kind of fix, not working example: the links in the sidebar permalink section fail to get the logmein treatment. I'll take a closer look at post preview iframes to see what we can do with postMessage for that.

Reply:
Added premapping of unmapped_url's onto their mapped domains.

Todo: postmessage for editor or etk fix.

Reply:
Update: "View Site" link on theme page now caught and replaced.

Reply:
Update: "Visit Site" in page/post preview is updated on right click but not left click.

Reply:
Tackling this in a follow up.